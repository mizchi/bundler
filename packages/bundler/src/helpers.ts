// TODO: publicPath
export function filepathToFlatSymbol(filepath: string, publicPath = "/") {
  return publicPath + filepath.replace(/\//g, "_$_");
}
