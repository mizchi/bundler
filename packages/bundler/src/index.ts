import { Bundler } from "./bundler";

export function bundle(
  files: { [k: string]: string },
  entry: string,
  options: {
    optimize?: boolean;
    exposeToGlobal?: string | null;
    publicPath?: string;
  } = {}
) {
  const bundler = new Bundler(files);
  return bundler.bundle(entry, options);
}

export { Bundler };
