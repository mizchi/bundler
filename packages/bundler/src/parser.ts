import { Ast } from "./types";
import { parse as parseAsBabel } from "@babel/parser";

export function parse(code: string, filepath: string): Ast {
  return parseAsBabel(code, {
    sourceFilename: filepath,
    sourceType: "module",
    plugins: ["typescript", "jsx"],
  });
}
