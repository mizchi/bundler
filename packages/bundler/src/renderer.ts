import type { AnalyzedChunk, Ast } from "./types";

import path from "path";
import { transformToEntryRunner, transformToModuleRunner } from "./transformer";
import generateAsBabel from "@babel/generator";

export function render(
  entryPath: string,
  chunks: AnalyzedChunk[],
  opts: {
    exposeToGlobal: string | null;
    transformDynamicImport: boolean;
    publicPath?: string;
  }
) {
  const moduleCodes: string[] = [];
  const basepath = path.dirname(entryPath);
  const entryModule = chunks.find((x) => x.filepath === entryPath)!;
  const moduleChunks = chunks.filter((x) => x.filepath !== entryPath);

  for (const chunk of moduleChunks) {
    const basepath = path.dirname(chunk.filepath);
    const runnerAst = transformToModuleRunner(chunk.ast, basepath, {
      transformDynamicImport: opts.transformDynamicImport,
      publicPath: opts.publicPath ?? "/",
    });
    const moduleCode = generate(runnerAst);
    moduleCodes.push(
      `"${chunk.filepath}": (_$_exports) => {${moduleCode}\nreturn _$_exports}`
    );
  }

  const code = "{" + moduleCodes.join(",") + "}";

  let additianalCode = "";
  if (opts.exposeToGlobal) {
    additianalCode += `/* Expose import */ globalThis.${opts.exposeToGlobal} = {import: _$_import};`;
  }

  const entryCode = generate(
    transformToEntryRunner(entryModule.ast, basepath, {
      transformDynamicImport: opts.transformDynamicImport,
      publicPath: opts.publicPath ?? "/",
    })
  );
  return runnerTemplate(code, additianalCode, entryCode);
}

export function generate(ast: Ast): string {
  const gen = generateAsBabel(ast);
  return gen.code;
}

const runnerTemplate = (
  moduleCodes: string,
  additianalCode: string,
  entryCode: string
) => `// @mizchi/bundler generate
const _$_exported = {};
const _$_import = (id) => _$_exported[id] || _$_modules[id](_$_exported[id] = {});
const _$_modules = ${moduleCodes};

${additianalCode};

// -- entry --

${entryCode};

`;
