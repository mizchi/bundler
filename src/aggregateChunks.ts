import { ModulesMap } from "./types";
import type { AnalyzedChunk } from "./types";

export function aggregateChunks(modulesMap: ModulesMap, entryPath: string) {
  const entryMod = modulesMap.get(entryPath)!;
  const chunks: AnalyzedChunk[] = [];
  _aggregate(entryMod);
  return chunks;

  function _aggregate(mod: AnalyzedChunk) {
    if (chunks.find((x) => x.filepath === mod.filepath)) {
      return chunks;
    }

    for (const imp of mod.imports) {
      if (chunks.find((x) => x.filepath === imp.filepath)) {
        continue;
      }
      const child = modulesMap.get(imp.filepath)!;
      _aggregate(child);
    }
    chunks.push(mod);
    return chunks;
  }
}
