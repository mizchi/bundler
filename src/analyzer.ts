import * as t from "@babel/types";
import type {
  Export,
  Import,
  Specifier,
  DynamicImport,
  Analyzed,
  WorkerImport,
  Ast,
  AstNode,
} from "./types";

import path from "path";
import traverse from "@babel/traverse";

export function analyzeModule(ast: Ast, basepath: string): Analyzed {
  let imports: Import[] = [];
  let exports: Export[] = [];
  let dynamicImports: DynamicImport[] = [];
  let workerImports: WorkerImport[] = [];

  const refs = getGlobalsWithoutImports(ast);
  // console.log("globals", refs);

  traverse(ast, {
    // detect dynamic import
    CallExpression(nodePath) {
      // console.log("CallExpression", nodePath.node);
      if (nodePath.node.callee.type == "Import") {
        const arg = nodePath.node.arguments[0];
        if (arg.type === "StringLiteral") {
          const absPath = path.join(basepath, arg.value);
          dynamicImports.push({
            filepath: absPath,
          });
        }
      }
    },
    ImportDeclaration(nodePath) {
      const target = nodePath.node.source.value;
      const absPath = path.join(basepath, target);
      const specifiers: Specifier[] = [];

      for (const specifier of nodePath.node.specifiers) {
        const used = refs.includes(specifier.local.name);
        if (specifier.type === "ImportSpecifier") {
          specifiers.push({
            localName: specifier.local.name,
            importedName: specifier.imported.name,
            used,
          });
        }
        if (specifier.type === "ImportDefaultSpecifier") {
          specifiers.push({
            localName: specifier.local.name,
            importedName: "default",
            used,
          });
        }
      }

      imports.push({
        filepath: absPath,
        specifiers: specifiers,
      });
    },
    ExportDeclaration(nodePath) {
      if (nodePath.node.type === "ExportNamedDeclaration") {
        const decl = nodePath.node.declaration.declarations[0];
        const pure = isPureNode(decl.init);
        exports.push({ exportedName: decl.id.name, pure });
      }
    },
  });
  return {
    imports,
    exports,
    dynamicImports,
    workerImports,
    pure: isPureProgram(ast),
  };
}

export function getGlobalsWithoutImports(ast: Ast) {
  let ids: string[] = [];
  const astWithoutImports = {
    ...ast,
    program: {
      ...ast.program,
      body: ast.program.body.filter((x) => x.type !== "ImportDeclaration"),
    },
  };
  traverse(astWithoutImports, {
    Program(path) {
      // @ts-ignore
      ids = Object.keys(path.scope.globals);
    },
  });
  return ids;
}

const WHITELIST_NODE_TYPE: string[] = [
  // Statement
  "ImportDeclaration",
  "FunctionDeclaration",
  "EmptyStatement",
  "DebuggerStatement",
  "ClassDeclaration",
  "EmptyStatement",

  // Expression
  "ArrowFunctionExpression",
  "FunctionExpression",
  "ClassExpression",
  "Identifier",

  // Literal
  "Literal",
  "StringLiteral",
  "BooleanLiteral",
  "NumericLiteral",
  "Literal",
];

function isPureNodeType(type: string): boolean {
  return WHITELIST_NODE_TYPE.includes(type);
}

export function isPureNode(node: AstNode) {
  switch (node.type) {
    case "VariableDeclaration": {
      for (const declaration of node.declarations) {
        if (declaration.init) {
          if (!isPureNode(declaration.init)) {
            return false;
          }
        }
        continue;
      }
      return true;
    }
    case "ObjectExpression": {
      for (const prop of node.properties) {
        if (prop.type === "SpreadElement") {
          if (!isPureNode(prop.argument)) {
            return false;
          }
          continue;
        }
        if (prop.computed) {
          if (!isPureNode(prop.key)) {
            return false;
          }
        }
        if (prop.type === "ObjectProperty") {
          if (!isPureNode(prop.value)) {
            return false;
          }
        }
      }
      return true;
    }

    case "ArrayExpression": {
      for (const element of node.elements) {
        // TODO: [[]]
        if (element && !isPureNode(element)) {
          return false;
        }
      }
      return true;
    }
    case "ExpressionStatement": {
      if (!isPureNode(node.expression)) {
        return false;
      }
      return true;
    }
    case "ExportNamedDeclaration": {
      if (!isPureNode(node.declaration)) {
        return false;
      }
      return true;
    }
    case "ExportDefaultDeclaration": {
      if (!isPureNode(node.declaration)) {
        return false;
      }
      return true;
    }
    default: {
      if (!isPureNodeType(node.type)) {
        return false;
      }
      return true;
    }
  }
}

export function isPureProgram(parsed: Ast): boolean {
  for (const stmt of parsed.program.body) {
    if (!isPureNode(stmt)) {
      return false;
    }
  }
  return true;
}
