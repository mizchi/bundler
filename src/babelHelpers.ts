import generateAsBabel from "@babel/generator";
import { parse as parseAsBabel } from "@babel/parser";
import type { Program } from "@babel/types";

export function generateRuntimeCode(ast: Program): string {
  const gen = generateAsBabel(ast);
  return gen.code;
}

export function parse(code: string, filepath: string): Program {
  return (parseAsBabel(code, {
    sourceFilename: filepath,
    sourceType: "module",
    plugins: ["typescript", "jsx"],
  }) as any) as Program;
}
