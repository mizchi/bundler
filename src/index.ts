import type { Program } from "@babel/types";
import type { IPromisesAPI } from "memfs/lib/promises";
import type {
  BundleOptions,
  InternalOptions,
  ParsedModule,
  ModulesMap,
} from "./types";

import path from "path";
import { generate } from "./generator";
import { parse } from "./parser";
import { analyzeScope } from "./analyzer";
import { transformToEntry, transformToRunnerModule } from "./transformer";
import { treeshake } from "./treeshaker";
import { createMemoryFs, readFile } from "./memfsHelpers";
import { isPure } from "./sideEffect";

export class Bundler {
  private modulesMap = new Map<string, ParsedModule>();
  public fs: IPromisesAPI;

  constructor(public files: { [k: string]: string }) {
    this.fs = createMemoryFs(files);
  }
  public async bundle(
    entry: string,
    { exposeToGlobal = null, preserveExternalImport = true }: BundleOptions = {}
  ) {
    const internalOptions: InternalOptions = {
      exposeToGlobal: exposeToGlobal,
      preserveExternalImport,
    };
    await this.addModule(entry);
    const ctx = new BuildContext(entry, this.modulesMap);
    return await ctx.emit(internalOptions);
  }

  public async updateModule(filepath: string, nextContent: string) {
    await this.fs.writeFile(filepath, nextContent);
    this.modulesMap.delete(filepath);
    this.addModule(filepath);
  }

  // TODO: need this?
  async deleteRecursive(filepath: string) {
    const cached = this.modulesMap.get(filepath)!;
    if (cached) {
      for (const i of cached.imports) {
        this.deleteRecursive(i.filepath);
        this.modulesMap.delete(i.filepath);
      }
    }
  }

  private async addModule(filepath: string): Promise<void> {
    if (this.modulesMap.has(filepath)) {
      return;
    }

    const basepath = path.dirname(filepath);

    const raw = await readFile(this.fs, filepath);
    const ast = parse(raw, filepath);

    // extract before transform
    const { imports, exports } = analyzeScope(ast, basepath);
    const hasSideEffect = !isPure(ast);

    // console.log("[addModule]", filepath, imports, exports);
    this.modulesMap.set(filepath, {
      raw,
      filepath,
      ast,
      imports,
      exports,
      hasSideEffect,
    });

    // console.log("used", filepath, JSON.stringify(imports, null, 2));
    for (const i of imports) {
      await this.addModule(i.filepath);
    }
  }
}

type BuildContextModulesList = Array<
  ParsedModule & {
    filepath: string;
    treeshaked: Program;
    importedBy: Array<{ filepath: string; importedName: string[] }>;
  }
>;

class BuildContext {
  ctxModulesMap: BuildContextModulesList = [];
  constructor(private entryPath: string, private modulesMap: ModulesMap) {}

  public async emit({ exposeToGlobal }: InternalOptions) {
    this.ctxModulesMap = [];
    const entryMod = this.modulesMap.get(this.entryPath)!;
    const basepath = path.dirname(entryMod.filepath);

    const treeshaked = treeshake(
      entryMod.ast,
      this.entryPath,
      [],
      this.modulesMap
    );

    const analyzed = analyzeScope(treeshaked, basepath);

    const mod = {
      ...entryMod,
      ...analyzed,
      importedBy: [],
      treeshaked,
    };
    // this.ctxModulesMap.push(mod);

    let additianalCode = "";
    if (exposeToGlobal) {
      additianalCode += `/* Expose import */ globalThis.${exposeToGlobal} = {import: _$_import};`;
    }
    this.bundleRecursively(mod);
    const moduleCodes: string[] = [];
    for (const i of this.ctxModulesMap.filter(
      (x) => x.filepath !== this.entryPath
    )) {
      const basepath = path.dirname(i.filepath);
      const runnerAst = transformToRunnerModule(i.treeshaked, basepath);
      const moduleCode = generate(runnerAst);
      moduleCodes.push(
        `"${i.filepath}": (_$_exports) => {${moduleCode}\nreturn _$_exports}`
      );
    }

    const code = "{" + moduleCodes.join(",") + "}";

    const runner = transformToEntry(treeshaked, basepath);
    const entryCode = generate(runner);
    return runnerTemplate(code, additianalCode, entryCode);
  }

  bundleRecursively(mod: ParsedModule) {
    if (this.ctxModulesMap.find((x) => x.filepath === mod.filepath)) {
      return;
    }
    const shaked = treeshake(mod.ast, mod.filepath, [], this.modulesMap);
    const basepath = path.dirname(mod.filepath);
    const analyzed = analyzeScope(shaked, basepath);

    this.ctxModulesMap.push({
      ...mod,
      ...analyzed,
      importedBy: [],
      filepath: mod.filepath,
      treeshaked: treeshake(mod.ast, mod.filepath, [], this.modulesMap),
    });

    for (const imp of analyzed.imports) {
      if (this.ctxModulesMap.find((x) => x.filepath === imp.filepath)) {
        continue;
      }
      const child = this.modulesMap.get(imp.filepath)!;
      this.bundleRecursively(child);
    }
  }
}

const runnerTemplate = (
  moduleCodes: string,
  additianalCode: string,
  entryCode: string
) => `// minibundle generate
const _$_exported = {};
const _$_import = (id) => _$_exported[id] || _$_modules[id](_$_exported[id] = {});

const _$_modules = ${moduleCodes};

${additianalCode};

// -- entry --

${entryCode};

`;
