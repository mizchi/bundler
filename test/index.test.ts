import { Bundler } from "./../src/index";
import { format } from "prettier";
import assert from "assert";

// runtime
const files = {
  "/bar.js": `
import foo from "./foo.js";
console.log("eval bar once")
export default "bar$" + foo
`,
  "/foo.js": `
console.log("eval foo once")
export default "foo$default";
export const b = "b";
export const a = "a";
`,
  "/index.js": `
import foo, {a, b as c} from "./foo.js";
import bar from "./bar.js";

export const x: number = c;
export default 1;

console.log(foo, bar);
`,
};

test("bundle", async () => {
  const bundler = new Bundler(files);
  const built = await bundler.bundle("/index.js", {});
  expect(format(built, { parser: "babel" })).toMatchSnapshot();

  const built$foo = await bundler.bundle("/foo.js", {});
  expect(format(built$foo, { parser: "babel" })).toMatchSnapshot();

  const built$bar = await bundler.bundle("/bar.js", {});
  expect(format(built$bar, { parser: "babel" })).toMatchSnapshot();
});

test(`export {x, y as z} from "./m.js"`, async () => {
  const bundler = new Bundler({
    "/up.js": "export const up = 1; export const down = 2;",
    "/m.js": `export {up as a, down} from "./up.js"`,
    "/index.js": `
    import { a as b } from "./m.js";
    globalThis.__export_b = b;
    `,
  });
  const built = await bundler.bundle("/index.js", { optimize: false });
  // console.log(format(built, { parser: "babel" }));
  expect(format(built, { parser: "babel" })).toMatchSnapshot();
  eval(built);
  // @ts-ignore
  assert.equal(globalThis.__export_b, 1);
});

test(`entry: export {default as b} from "./m.js"`, async () => {
  const bundler = new Bundler({
    "/up.js": "export default 1;",
    "/m.js": `export {default as a} from "./up.js"`,
    "/index.js": `
    export { a as b } from "./m.js";
    `,
  });
  const built = await bundler.bundle("/index.js", {});
  console.log(format(built, { parser: "babel" }));
  expect(format(built, { parser: "babel" })).toMatchSnapshot();
});
