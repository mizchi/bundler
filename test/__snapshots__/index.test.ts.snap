// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle 1`] = `
"// minibundle generate
const _$_exported = {};
const _$_import = (id) =>
  _$_exported[id] || _$_modules[id]((_$_exported[id] = {}));

const _$_modules = {
  \\"/foo.js\\": (_$_exports) => {
    console.log(\\"eval foo once\\");
    _$_exports.default = \\"foo$default\\";
    _$_exports.b = \\"b\\";
    _$_exports.a = \\"a\\";
    return _$_exports;
  },
  \\"/bar.js\\": (_$_exports) => {
    const { default: foo } = _$_import(\\"/foo.js\\");

    console.log(\\"eval bar once\\");
    _$_exports.default = \\"bar$\\" + foo;
    return _$_exports;
  },
};

// -- entry --

const { default: foo, a: a, b: c } = _$_import(\\"/foo.js\\");

const { default: bar } = _$_import(\\"/bar.js\\");

export const x: number = c;
export default 1;
console.log(foo, bar);
"
`;

exports[`bundle 2`] = `
"// minibundle generate
const _$_exported = {};
const _$_import = (id) =>
  _$_exported[id] || _$_modules[id]((_$_exported[id] = {}));

const _$_modules = {};

// -- entry --

console.log(\\"eval foo once\\");
export default \\"foo$default\\";
export const b = \\"b\\";
export const a = \\"a\\";
"
`;

exports[`bundle 3`] = `
"// minibundle generate
const _$_exported = {};
const _$_import = (id) =>
  _$_exported[id] || _$_modules[id]((_$_exported[id] = {}));

const _$_modules = {
  \\"/foo.js\\": (_$_exports) => {
    console.log(\\"eval foo once\\");
    _$_exports.default = \\"foo$default\\";
    _$_exports.b = \\"b\\";
    _$_exports.a = \\"a\\";
    return _$_exports;
  },
};

// -- entry --

const { default: foo } = _$_import(\\"/foo.js\\");

console.log(\\"eval bar once\\");
export default \\"bar$\\" + foo;
"
`;
